feat: gettext/babel for translation

ssiyad opened the pull request:
### Spec
- Locale dir: `app_dir/app/locale/`. eg: `frappe/frappe/locale`
- Template file: `locale_dir/main.pot`
- Locale PO file: `locale_dir/locale/LC_MESSAGES/messages.po`
- Locale MO file: `locale_dir/locale/LC_MESSAGES/messages.mo`

### New bench commands
- `generate-pot`
- `compile-translation`
- `migrate-translation`
- `update-po`
- `new-po`

### Removed bench commands
- `update-translations`
- `new-language`
- `build-message-files`
- `get-untranslated`
- `import-translations`

### How to migrate

```bash
# generate template file
bench generate-pot --app my_app

# generate .po file for all languages
bench migrate-translation --app my_app

# compile .po files
bench compile-translation --app my_app

# remove old translations
rm -r apps/my_app/my_app/translations
```

### Extractors
- [x] Python
- [x] DocType JSON
- [x] JavaScript
- [x] Page
- [x] Workspace
- [ ] HTML Templates
- [ ] Report (Covered in `Javascript`?)
- [ ] Navbar (`standard_navbar_items`, `standard_help_items`)
- [ ] Included files (?)

### Todo
- [ ] migration of sub-languages, for example `es_MX`. Maybe copy `es.po`, then update from `es-mx.csv`.
- [ ] apps can provide their own extractors, for files not covered by frappe's standard extractors
- [ ] generate the `.mo` files only locally and not include them in the repository (similar to JS build process).
- [ ] Maybe the folder structure can be flattened, removing the extra `<lang>/LC_MESSAGES` folders? Not sure what was the original intention here.

Resolves #15325

ssiyad committed:
feat(translation): base version

- use gettext
- convert existing csv to po

ssiyad committed:
fix(translation): escape '%'

ssiyad committed:
feat(i18n): engine

ssiyad committed:
feat(i18n): get messages for boot

ssiyad committed:
feat(i18n): get user translations

ssiyad committed:
refactor(i18n): use constants

ssiyad committed:
Merge branch 'develop' into feat/translate_gettext

ssiyad committed:
refactor(i18n): move to `translation.py`

ssiyad committed:
chore(translate): format

ssiyad committed:
doc(translate): comment translation function

ssiyad committed:
refactor(translate): change imports to `translate.py`

ssiyad committed:
feat(commands): generate pot file (base, needs upgrade)

ssiyad committed:
feat(commands): migrate csv to po

ssiyad committed:
refactor: remove `frappe/i18n`

ssiyad committed:
fix(commands): `generate_pot` -> `generate-pot`

ssiyad committed:
feat(babel): (our) context aware python extractor

thanks to @barredterra
ref: https://github.com/frappe/frappe/pull/19238

ssiyad committed:
feat(commands): compile translations. `po` -> `mo`

ssiyad committed:
feat(commands): update `po`: sync with `pot` file

ssiyad committed:
feat(command): new po file

ssiyad committed:
refactor!(command/translate): remove `update-translations`

ssiyad committed:
refactor!(command/translate): remove `build-message-files`

ssiyad committed:
refactor!(command/translate): remove `import-translations`

ssiyad committed:
refactor!(command/translate): remove `new-language`

ssiyad committed:
refactor!(command/translate): remove `get-untranslated`

ssiyad committed:
fix: remove unused commands

ssiyad committed:
feat: boot messages: get from hooks

ssiyad committed:
fix(translation): generate: do not skip tuple

ssiyad committed:
refactor!(translate): remove unused methods

ssiyad committed:
feat(translation): babel extractor for doctype json

copied from https://github.com/frappe/frappe/pull/19238
credits @barredterra

barredterra commented on frappe/translate.py:
```diff
@@ -1229,6 +688,85 @@ def get_messages(language, start=0, page_length=100, search_text=""):
+					continue
+
+				messages.extend(
+					(option, f"Option for a {fieldtype} field in DocType '{doctype}'")
```
```suggestion
					(option, f"Option for a Select field in DocType '{doctype}'")
```

barredterra commented on frappe/translate.py:
```diff
@@ -1229,6 +688,85 @@ def get_messages(language, start=0, page_length=100, search_text=""):
+				)
+			elif fieldtype == "HTML":
+				messages.append(
+					(message, f"Content of an {fieldtype} field in DocType '{doctype}'")
```
```suggestion
					(message, f"Content of an HTML field in DocType '{doctype}'")
```

barredterra reviewed (changes requested):
Don't know what I was thinking here... using format string for invariants is unnecessary and harder to read.

ankush commented:
@ssiyad migrate all CSV files and replace with .po files? so we can actually test this too.

ssiyad committed:
unwanted use of format strings

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

ssiyad committed:
unwanted use of format string

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

barredterra commented:
Just came across this: apps should be able to provide their own extractors / arbitrary translatable strings. For example, the ERPNext setup wizard installs lots of fixtures that need to be translated (UOMs, Lead Types, etc.). There should be some way for ERPNext to provide these strings to the potfile generation.

Let me know if that's something that could go into this PR or if I should raise a separate feature request.

By the way, great effort so far!

ssiyad commented:
keep-alive

barredterra commented:
@ssiyad please don't give up on this 🙏🏼 I know you can do it 👍🏼

ssiyad committed:
chore: git index `.pot`, `.po` and `.mo` files

ssiyad committed:
chore: remove `./translations`

ssiyad committed:
feat: extractor for javascript

ssiyad committed:
feat(translation): extract from jinja2 pages

use `jinja2.ext:extract_babel`

barredterra committed:
Merge remote-tracking branch 'upstream/develop' into feat/translate_gettext

barredterra committed:
reafctor: bench translation commands

barredterra committed:
feat: use metadata from app hooks

barredterra committed:
feat: include message metadata in potfile

barredterra reviewed (commented):
@ssiyad looks pretty good already! I took the liberty to push a couple of small fixes (and messed up a conventional commit – sorry for that).

- Refactored a common method in translation commands
- Use header metadata from app hooks instead of hardcoding these values
- Include message metadata in the potfile (filenname, fileno, comments, context)

barredterra committed:
feat: option to compile a specific locale only

barredterra committed:
feat: various changes to translate.py

- Feat: add locale parameter to migrate-translation
- Refactor: helper methods for common actions
- Fix: migrate-translation produces full po file

barredterra committed:
fix: underscore in locale

barredterra committed:
refactor: extract get_locales

barredterra committed:
chore: regenerate message files

barredterra commented:
- Some more refactoring
- Added `--locale` option for generating files for a single locale (useful for faster testing)
- `migrate-translation`: start with the app's potfile, add existing translations wherever possible. This results in a much larger file with much more translations, but it has to be like this.
- `update-translations`: use `catalog.update(template)`, which works as expected with a complete po file generated by `migrate-translation`

Given the sitze of the translation files, I would suggest to generate the `.mo` files only locally and not include them in the repository.

ssiyad commented:
@barredterra Thanks for the changes!

> Given the size of the translation files, I would suggest to generate the .mo files only locally and not include them in the repository.

Makes sense. Lets keep it that way. But then we will have to compile it on bench start (or install) with `compile-translation`

I think translation from existing catalogs work fine. What we have to ensure is that every translatable string is cataloged. I kept a check list in the opening comment. Can you check if anything is missing from that?

barredterra committed:
feat: add workspace extractor

barredterra commented:
> What we have to ensure is that every translatable string is cataloged.

Added an extractor for workspaces. HTML templates and navbar are not yet covered.

Unfortunately I won't be able to contribute more this week.

ankush commented:
@ssiyad @barredterra 

is it possible to broadly list down what's pending in this for merge?

barredterra commented:
@ankush apart from the todos in the PR description?

ankush commented:
@barredterra yeah apart from description just in case anything else was discovered.

barredterra commented:
@ankush when I last worked on this, I added everything in the todo and missing extractors lists. No additional insights from me.

gavindsouza commented on frappe/locale/am/LC_MESSAGES/messages.mo:
We can get rid of `MO` from tracking as they are generated using `PO` files?

barredterra replied:
Yes, I suggested the same in https://github.com/frappe/frappe/pull/19655#issuecomment-1485881357

barredterra replied:
Didn't remove yet because we'd first have to include them in the build process

barredterra committed:
chore: re-add csv files

barredterra committed:
chore: remove pot/po/mo files

barredterra commented:
I temporarily reverted the generated bulk changes, in order to be able to view the diff without crashing the browser.

surajshetty3416 commented:
Closing due to inactivity. 

@ssiyad maybe list out things that are pending so that others and pick it up if you don't have time to continue this.

barredterra commented:
@surajshetty3416 I've already taken this over in https://github.com/frappe/frappe/pull/22732
